Support vncmanager 3

michalsrb opened the pull request:
Add support for vncmanager mode to Remote. fate#319319

Remote module can now set up 3 VNC modes: disabled, xinetd, vncmanager
* disabled
  * vncmanager systemd service is disabled
  * vnc1 and vnchttpd1 xinetd services are disabled
* xinetd
  * vncmanager systemd service is disabled
  * xinetd systemd service is enabled
  * vnc1 and vnchttpd1 xinetd services are enabled
* vncmanager
  * vncmanager systemd service is enabled
  * xinetd systemd service is enabled
  * vnchttpd1 xinetd service is enabled but vnc1 is not

Calls to Remote.Enable and Remote.Disable set xinetd and disabled modes, so
users of the old API can keep using it.

mchf commented on src/include/network/remote/dialogs.rb:
```diff
@@ -60,17 +60,25 @@ def RemoteMainDialog
               Id(:disallow),
               _("&Do Not Allow Remote Administration"),
-              Remote.IsDisabled
+              Remote.Mode == :disabled
```
Basically all these ```Remote.Mode == ...``` checks breaks encapsulation. I'd stay with ```Remote.IsDisabled``` or better (according our new coding style) ```Remote.disabled?```

mchf commented on src/include/network/remote/dialogs.rb:
```diff
@@ -91,8 +99,7 @@ def RemoteMainDialog
               "client, such as krdc (connect to <tt>&lt;hostname&gt;:%1</tt>), or\n" \
-              "a Java-capable Web browser (connect to <tt>http://&lt;hostname&gt;:%2/</tt>).\n" \
-              "This form of remote administration is less secure than using SSH.</p>\n"
+              "a Java-capable Web browser (connect to <tt>https://&lt;hostname&gt;:%2/</tt>).\n"
```
I think it is not a good idea to change text which is subject of translation in this phase.

mchf commented on src/include/network/remote/dialogs.rb:
```diff
@@ -142,12 +149,15 @@ def RemoteMainDialog
 
-        allowed = Convert.to_boolean(UI.QueryWidget(Id(:allow), :Value))
+        allowed_xinetd = Convert.to_boolean(UI.QueryWidget(Id(:allow_xinetd), :Value))
+        allowed_vncmanager = Convert.to_boolean(UI.QueryWidget(Id(:allow_vncmanager), :Value))
```
we try to avoid using helpers which were introduced during ycp -> ruby conversion in the new code. You can use just ```UI.QueryWidget(...)``` here

mchf commented on src/include/network/remote/dialogs.rb:
```diff
@@ -142,12 +149,15 @@ def RemoteMainDialog
+          Remote.SetMode :vncmanager
         else
-          Remote.Disable
+          Remote.SetMode :disabled
```
similar as above. Pls stay with ```Remote.Disable``` or better ```Remote.disable```. Similar for other cases.

mchf commented on src/modules/Remote.rb:
```diff
@@ -70,26 +72,38 @@ def main
+    # Retrieves the current remote administration mode
+    def Mode
+      @mode
+    end
```
```attr_reader :mode```

mchf commented on src/modules/Remote.rb:
```diff
@@ -70,26 +72,38 @@ def main
     def Disable
-      @allow_administration = false
+      SetMode(:disabled)
 
```
So, you kept ```IsDisabled``` and ```Disable``` methods. That's another reason for using them ... see above.

mchf commented on src/modules/Remote.rb:
```diff
@@ -135,18 +149,25 @@ def Read
-        s == "vnc1" || s == "vnchttpd1"
+
+      xinetd_vnc1_enabled = false
+      xinetd_vnchttp1_enabled = false
```
These assignments seems to be useless. Default value for uninitialized variables is ```nil``` and it is treated as ```false``` in ruby.

michalsrb replied:
I am not so fluent in ruby, but the reason was to declare the current block so they can be accessed later. Without these lines they exist only in the inner do-end block, and attempt to access them later gives:

`NameError: undefined local variable or method `xinetd_vnc1_enabled'` at src/modules/Remote.rb:149

mchf replied:
right. I've missed that.

mchf commented on src/modules/Remote.rb:
```diff
@@ -135,18 +149,25 @@ def Read
+      vncmanager = Service.Enabled(VNCMANAGER_SERVICE)
 
-      @allow_administration = xdm && dm_ra && xinetd && vnc
+      @mode = :disabled
```
```Disable()```

mchf commented on src/modules/Remote.rb:
```diff
@@ -179,7 +200,8 @@ def WriteXinetd
         Ops.set(m, "changed", true)
-        Ops.set(m, "enabled", @allow_administration)
+        Ops.set(m, "enabled", @mode == :xinetd) if s == "vnc1"
+        Ops.set(m, "enabled", @mode != :disabled) if s == "vnchttpd1"
```
Similar to ```Convert``` a few comments above. Pls use just ```m["enabled"] = ...``` It helps us to clean the code ;-)

mchf commented on src/modules/Remote.rb:
```diff
@@ -235,25 +257,36 @@ def Write
+        services = [
+          [XINETD_SERVICE, true],
+          [XDM_SERVICE_NAME, true],
+          [VNCMANAGER_SERVICE, @mode == :vncmanager]
```
You sometime use ```Mode() == :vncmanager``` and sometime ```@mode == :vncmanager```. It is inconsistent and hard to maintain in case of further changes. Pls add ```Remote#vncmanager?``` & co (or something like that, naming methods is not my favorite thing to do).

mchf commented:
+10 for writing tests. Thank you.

michalsrb commented:
Thank you for the review, I will change the new code to use the new style.

But I will keep the original API (like Remote.IsEnabled, Remote.Enable, etc) the way it is - it seem to be used in other places in yast-network and also yast-installation, so changing them would require bigger rewrites.

michalsrb commented:
Please check the updated version. I think I addressed all your comments.

I know the new method "Remote#UsesVncManager" is using incorrect naming convention, but all the other methods in the class have that kind of names. I can't rename them all now and it just looked too weird mixing them together.

mchf commented on src/modules/Remote.rb:
```diff
@@ -72,38 +74,46 @@ def main
     # Enables remote administration.
-    def Enable
-      @allow_administration = true
+    def Enable(use_vncmanager = false)
```
this boolean option is bad. It makes code hardly readable and nonintuitive as you can see few lines bellow (https://github.com/yast/yast-network/pull/401/files#diff-3e7dc301d1e2f6c63659b651f1c3ed2dR161)

I'd stay with ```Enable``` with no option for xinetd to stay backward compatible and introduce e.g. ```enable_with_persistent_sessions``` for vncmanager.

michalsrb replied:
Alright, I'll change that.

mchf commented on src/include/network/remote/dialogs.rb:
```diff
@@ -60,9 +60,17 @@ def RemoteMainDialog
-              Remote.IsEnabled
+              Id(:allow_with_vncmanager),
+              _("&Allow Remote Administration With Session Management"),
+              Remote.IsEnabled && Remote.UsesVncManager
```
I think that ```Remote.IsEnabled``` can be checked in ```Remote.UsesVncManager```

michalsrb replied:
Sure it can. I made it like this for the symmetry with the next radio button.

mchf commented on src/modules/Remote.rb:
```diff
@@ -135,18 +145,30 @@ def Read
-      @allow_administration = xdm && dm_ra && xinetd && vnc
+      vncmanager = Service.Enabled(VNCMANAGER_SERVICE)
+
+      if xdm && dm_ra && xinetd && xinetd_vnchttp1_enabled
```
it would be nice to hide this long condition into a private helper. So, I'd do something like: ```if remote_access_requested```

michalsrb replied:
I am not sure what you mean by a private helper. That expression is composed of local variables and in that method exists exactly once. How can it be meaningfully extracted into a helper?

kobliha commented on src/modules/Remote.rb:
```diff
@@ -135,18 +152,30 @@ def Read
+      vncmanager = Service.Enabled(VNCMANAGER_SERVICE)
 
-      @allow_administration = xdm && dm_ra && xinetd && vnc
+      if xdm && dm_ra && xinetd && xinetd_vnchttp1_enabled
```
```ruby
if something?
   ...
end

# Returns X whether something
def something?
    xdm && dm_ra && xinetd && xinetd_vnchttp1_enabled
end
```

michalsrb replied:
But `xdm`, `dm_ra`, `xinetd` and `xinetd_vnchttp1_enabled` are local variables existing only inside `Remote#Read`.

mchf replied:
I most probably still don't understand this ```if ... else``` construct correctly. What is expected combination of these flags when vncmanager should be running? According this condition it e.g. seems that xinet should be running.

michalsrb replied:
That is correct; xinetd is used even in vncmanager mode to serve the web page with VNC applet, but not to spawn the Xvnc processes (that is done by vncmanager instead). It is described in the pull request message, but I should have made this clearer.

So:

- Enabled (vncmanager mode)
`xdm && dm_ra && xinetd && xinetd_vnchttp1_enabled && !xinetd_vnc1_enabled && vncmanager`

- Enabled (xinetd mode)
`xdm && dm_ra && xinetd && xinetd_vnchttp1_enabled && xinetd_vnc1_enabled && !vncmanager`

- Disabled
any other combination

mchf replied:
thanks for repeating that ...

One more thing is unclear to me. Why is ```xinetd_vnchttp1_enabled``` needed when enabling vnc1 or vncmanager service? I'd think of it as independent / alternative to these services.

Sorry, that it takes it so long. I want to use this opportunity to make some dark corners more clear to me.

michalsrb replied:
The `vnchttp1` depends on the others. The `vnchttp1` is a simple web server that gives you a page with a java vnc viewer that connects to vnc server run by the `vnc1` service or `vncmanager`.

So running `vnchttp1` alone would be pointless. We could make `vnchttp1` optional when enabling the `vnc1` or `vncmanager`, but that would be another feature and another pull request.

mchf replied:
That's what I've somehow expected. But written this way it seems so that ```vnc1``` and / or ```vncmanager``` depends on ```vnchttp1```. So, when e.g. user edits ```/etc/xinetd.d/vnc``` manually and disables ```vnchttpd1```  and keeps ```vnc1``` enabled then the code will disable remote access completely. 

In general I don't see a point in putting ```vnchttpd1``` check into condition for starting ```vnc1``` or ```vncmanager``` there is no relation in this direction.

michalsrb replied:
It does behave the way you are describing. Which is the same way as it used to.

It we omit the `vnchttpd1` from this condition, it will get re-enabled after submitting the form.

I don't have any preferences either way, if you think it's better to omit it here, I will change it.

mchf replied:
I just wanted to clarify that I get it correctly (that the code really is so messy ;-)

I'm aware of re-enabling in write.

I'd definitely drop it from here and I'd be glad if you could create a fate request which resolves the write (or asks for separate handling of ```vnchttp1``` at all). Thank you for the patience.

michalsrb commented:
Changed the `Remote#Read` to ignore the state of the `vnchttpd1` service.

Created [fate#320881: Make VNC java applet service optional in YaST's VNC setup](https://fate.suse.com/320881).

mchf commented:
Thank you for the patch, for the patience and discussion. I'll add some necessary administrative stuff and merge it.

michalsrb commented:
Great, thank you too.

mchf commented:
Updated patch submitted, so closing this one.
